accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1461,3 +1461,319 @@ wrapper class is so huge that it is likely to be a major use case for
 extension types that they can allow us to use a built-in class as the
 representation, and still have a specialized interface&mdash;that is, an
 extension type.
+
+## Specification changes related to specific types
+
+Extension types can implement class types, even class types and their members,
+which are mentioned in the language specification.
+
+By disallowing extension members with the same names as the members of `Object`,
+we have reduced the problem to where the specification refers to members of
+specific interfaces. 
+The specification needs to be updated to account for the existence of extension
+types that are subtypes of those types, and which possibly shadow members from
+the interface with extension members.
+
+The specification, and feature specifications not yet included in the
+specification, may uses phrases like (taken from the original collection
+elements feature, so no longer direttly relevant):
+
+> 1. Evaluate the iterator expression to a value `sequence`.
+> 2. If `sequence` is not an instance of a class that implements `Iterable`,
+>    throw a dynamic exception.
+> 3. Evaluate `sequence.iterator` to a value `iterator`.
+
+Here the “Evaluate `sequence.iterator`” is imprecise, and really means “Invoke
+the instance member `iterator` on the object `sequence` (which must exist
+because of the second step). However, as *written*, it’s now ambiguous *which*
+`iterator` member to invoke, in case the static type of `sequence` is an
+extension type which implements `Iterable` and shadows `iterator`.
+
+**General Rule:** Where the existing language specification can invoke or access
+an extension member, it can also invoke an extension type member, and where it
+cannot invoke an extension member, it also will not invoke an extension type
+member.
+
+An example of the former is implicit `call` invocation: If `e1` in `e1(e2)` a
+static type which is not a function type, but which exposes a `call` method, the
+semantics is equivalent to that of `e1.call(e2)`. We apply this rule even the
+static type of `e1` itself has no `call` method, but an extension `call` method
+applies. If `e1` has an extension-type type, we check whether that has a `call`
+method, and not its representation type. There is no *type* check, only a
+*member* check.
+
+The latter case is usually signaled by the specification *requiring* that
+something implements a specific interface, and then invoking members of that
+interface on the value. We can generally treat those cases as if they first
+*up-cast* to the type they tested for, before invoking members, which ensure
+that they invoke instance members. 
```

Comment:
We generally need to distinguish between places where you *must* invoke an instance member, and where you *may* invoke an extension/extension type member.

In the former cases, we can distinguish between *always* invoking the instance member (which a type check has ensured is there), or refusing to run if the instance member is shadowed by an extension member.

In the latter cases, we may have rules that apply only if the invocation is an instance member invocation.

* The "must invoke instance member, must not be shadowed" is things like `const MyInt(1) + 1`, where `+` denotes an extension type member. That's a compile-time error. This couldn't happen with extension methods, since they wouldn't apply if there was an instance member.

* The "must, and always does invoke instance member, whether shadowed or not" is, as currently written, something like `for (var x in MyIterable())`. Here we check that `MyIterable` implements (or is assignable to) `Iterable<T>`. Then we invoke the instance `get iterator`, as if by first casting to `Iterable<T>` . We don't need to cast, we just make no assumptions about types other than it being an `Iterable<T>`, for example we don't check the actual type of the `Iterator<S>` that `iterator` provides, we just use it as an `Iterator<T>`. That's how code works today, since extension members wouldn't apply anyway, but extension type members might.

* The "may invoke extension members, but some rules only apply to instance methods" are rules like the special-ase number type inference, where `MyInt(1) + 1`, `MyInt` implements `int`, has static type `int` if `+` denotes an instance method, and whatever the return type of `MyInt.+` is if it denotes an extension type method. This couldn't happen with extension methods.

* And the "may invoke extension members, nothing special at all" includes the "implicit `call` invocation/tear-off". This is where we'd previously accept extension methods.


If we want to move `for (var x in e)` into the fourth category (from the second), we can. We need to be very careful about how we do it, so that we don't introduce unsoundness, and also don't introduce changes to how a normal `Iterable` is treated.

Something like:
> (.. try to infer an element type from `var x` or whatever that clause is, a type schemea `C`...)
> * Let `S` be the static type of `e` inferred with context type `Iterable<C>`.
> * If `S` is `dynamic`, let `S` be `Iterable<dynamic>`. _(And we'll do an implicit downcast later)_
> * If BOTTOM(`S`), let `E` be `Never`.
> * Otherwise if `S` implements `Iterable<T>` for some type `T`, let `E` be `T`.
> * Otherwise if `S` exposes an extension or extension-type getter named `iterator`:
>     * let *R* be the return type of the `S.iterator` getter.
>     * If *R* implements `Iterator<V>` for some `V`, let `E` be `V`.
>     * Otherwise if `R` has an instance, extension or extensions type method named `moveNext` which can be invoked with zero arguments,
>        and whose returns a

Reply:
A possible attempt to specify a more structural `for`/`in`:

Static semantics
> (.. try to infer an element type from `var x` or whatever that clause is, a type schemea `C`...)
> * Let `S` be the static type of `e` inferred with context type `Iterable<C>`.
> * If `S` is `dynamic`, let `S` be `Iterable<dynamic>`. _(And we'll do an implicit downcast later)_
> * If BOTTOM(`S`), let `E` be `Never`.
> * Otherwise if `S` implements `Iterable<T>` for some type `T`, let `E` be `T`.
> * Otherwise if `S` exposes an extension or extension-type getter named `iterator`:
>     * let *R* be the return type of the `S.iterator` getter.
>     * If *R* implements `Iterator<V>` for some `V`, let `E` be `V`.
>     * Otherwise if `R` has an instance, extension or extensions type method named `moveNext` 
>        which can be invoked with zero arguments, and whose return type is assignable to `bool`,
>        *and* `R` has an instance, extension or extension type getter named `current` with
>        return type `W`, then let `E` be `W`.
>    * otherwise a compile-time error occurs.
> * (... use `E` as the actual element type that we push back into `var x`. ...)
and runtime semantics:
> * Evaluate `e` to a value *v* with runtime type *O*.
> * If the static type of `e` is `dynamic`.
>     * If the runtime type of *v* is not a subtype of `Iterable<dynamic>`,  a runtime type error occurs.
>     * let *i* be the result of invoking the `iterator` getter on *v*.
>     * let *I* be `Iterator<dynamic>` _(aka. `Iterator<E>`)_.
> * Otherwise, if the static type of `e` implemented `Iterable<T>` , 
>     * let *i* be the result of invoking the `iterator` getter on `*v*
>     * let *I* be `Iterator<T>` _(aka. `Iterator<E>`)_.
> * Otherwise the static type of `e` has an extesion or extension type getter named `iterator`.
>     * let *i* be the result of invoking the denoted extension or extension type `iterator` getter
>        with *v* as receiver.
>     * if the denoted `iterator` getter's return type, `R`, implemented `Iterator<V>`,
>        let *I* be `Iteartor<V>` _(aka. `Iterator<E>`)_
>     * otherwise let `I` be `R`:
>  _(Then the iteration commences)_
> * While iterating, do:
>    * invoke the `I.moveNext` method with *i* as receiver, let *n* be the return value.
>    * if the runtime type of *n* is not `bool` a runtime error occurs _(possible downcast from `dynamic`)_.
>    * if *n* is false, we stop iterating, and the execution of the `for`/`in` loop completes.
>    * Otherwise *n* is true.
>    * Invoke the `I.current` getter with *i* as receiver, let *c* be the return value.
>    * If the initial part of the `for`/`in` is a declaration, 
>       * perform initialization of the loop declaration/pattern with *c* as matched value.
>    * otherwise assign *c* to the pre-existing loop variable.
>    * execute the loop body in a scope where the new variables of the loop declaration are in scope.
>    * Continue iterating from the top.

This is not the only, or necessarily the best, design. (And it still hides the complexity of choosing whether to invoke an extension/extension type member or an instance member by just saying "invoke `I.member`".

We could, possibly should, say that we give precedence to extension type overrides of `iterator`, `moveNext` and `current`, even when the type implements `Iterable`, but if we do that, we step out of the `Iterable` framework, and have to look at the actual return type of `current` to find the element type.
Even if the type implements `Iterable<num>`, it might have a `current` returning `MyNum`, which perhaps implements `double`. 
Today, we don't look that far down, we just say "implements `Iterable<E>`, OK, element type is `E`".

Also, the above does not allow iterating something that has an *instance* member named `get iterator`, but which doesn't implement `Iterable`. Why not?
Because we actually wanted to push people towards using `Iterable`. 
We could go back on that and go fully dynamic, say:

* If `S` is dynamic, let `E` be `dynamic`.
* else if BOTTOM(`S`), let `E` be `Never`
* else if `S` _has_ a getter named `Iterator` with return type `R`, and
   * `R` has a method named `moveNext` which can be called with zero arguments,
      and has a return type assignable to `bool`, and
   * `R` _has_ a getter named `current` with return type `T`,
   * then let `E` be `T`.
* else a compile-time error occurs.
* It's a compile-time error if `E` is not assignable to the loop variable.

and runtime behavior would be just like desugaring `labels: for (decl in e) body` to:
```dart
{
  S $tmp1 = e;
  R $tmp2 = $tmp.iterator;
  labels: while ($tmp2.moveNext()) {
    decl = $tmp2.current;
    { body }
  }
}
```
Here we don't try to do any kind of downcast from `dynamic`, we do dynamic invocations of `iterator`, `moveNext` and `current` (with predictably bad performance).

But you can iterate anything, whether it's an `Iterable` or not. The `Iterable` is there to provide all the *other* methods.